### DISCUSSION TITLE
물리주소와 논리주소, 메모리 보호 기법

### DISCUSSION THREAD

**Discussion by 3Juhwan** [2024-woowacourse-study/level-interview#264] [Computer Architecture]
.

**Comment by 3Juhwan** [comment]
## 물리주소와 논리주소

- 물리주소: 메모리 하드웨어가 사용하는 실제 주소이다. 
- 논리주소: CPU와 실행 중인 프로그램이 사용하는 주소로, 실행 중인 프로그램 각각에 0번지부터 시작되는 주소이다. 

## MMU

- MMU는 컴퓨터 시스템에서 메모리 관리와 주소 변환을 담당하는 하드웨어 컴포넌트다. 
- CPU가 프로그램을 실행하며 메모리에 접근하기 위해서 논리주소를 물리주소로 변환해야 한다. 
- 베이스 레지스터는 프로그램의 첫 물리 주소를 저장하고 있다. 
- 물리주소 = 논리주소 + 베이스 레지스터 값

## 메모리 보호 기법

- 프로세스의 독립적인 실행 공간을 확보하고, 다른 프로세스의 영역을 침범하지 않도록 보호해야 한다. 
- 프로세스가 사용할 수 있는 메모리 공간의 크기는 한계 레지스터(limit register)에 저장한다. 
- 프로세스가 사용할 수 있는 메모리 주소 공간은 [`베이스 레지스터`, `베이스 레지스터` + `한계 레지스터`] 이다. 
- 만약 CPU가 한계 레지스터보다 높은 논리 주소를 접근하려 하면 인터럽트(트랩)이 발생한다.

**Reply by 3Juhwan** [reply]
MMU를 사용하는 이유가 궁금해서 찾아본 내용

## 주소 바인딩 기법

**컴파일 타임 바인딩**
- 프로그램이 컴파일될 때 메모리 주소를 결정하는 방식이다. 
- 프로그램의 모든 주소를 컴파일 시점에 결정하므로, 프로그램이 메모리에 적재될 때 주소가 고정되어 있다. 
- 주소 결정이 간단하며, 프로그램의 실행이 빠르다. 
- 프로그램이 다른 메모리 위치에 적재되면 재컴파일이 필요하다.

**로드 타임 바인딩**
- 프로그램이 메모리에 적재될 때 주소를 결정하는 방식이다. 
- 컴파일 시에는 상대 주소를 사용하며, 프로그램이 메모리에 적재될 때 실제 주소로 변환된다.
- 주소 계산이 유연해져 재컴파일이 필요 없다. 
- 프로그램의 메모리 위치가 변경되면 주소를 다시 계산해야 하며, 위치 변경에 대한 지원이 부족하다. 
- 프로그램의 메모리 위치가 변경은 OS의 메모리 최적화 등의 상황을 의미한다. 

**실행 타임 바인딩**
- 프로그램이 실행될 때 주소를 결정하는 방식이다. 
- 주소 변환을 실행 시간에 처리하며, MMU(메모리 관리 장치)를 통해 논리적 주소와 물리적 주소 간의 변환이 이루어진다.
- 동적 메모리 관리와 위치 변경의 유연성을 제공한다.
- 주소 변환을 위해 MMU가 필요하며, 메모리 접근 시간이 2배로 늘어날 수 있다.

**Comment by le2sky** [comment]
**메모리 구성**

<p align = "center">
   <img height = 400 width = 300 src = "https://github.com/user-attachments/assets/45e40084-0cc4-4a5a-854d-5d785c2ec74f" />

각 프로세스는 별도의 메모리 공간이 존재한다. 이때 개별적인 메모리 공간을 분리하기 위해서 특정 프로세스만 접근할 수 있는합법적인 메모리 주소 영역을 설정하고 프로세스가 합법적인 영역만을 접근하도록 하는 기능이 필요하다. 이때 베이스 레지스터와 한계 레지스터를 사용한다. 

- 베이스 레지스터 : 가장 작은 합법적인 물리 메모리 주소
- 한계 레지스터 : 주어진 영역의 크기

---

**메모리 공간 보호**
<p align = "center">
   <img height = 300 width = 450 src = "https://github.com/user-attachments/assets/fed89934-3b45-44fe-906f-a2554edb3d0e" />
</p>

- 베이스와 한계 레지스터를 사용해 사용자 프로그램이 운영체제나 다른 사용자 코드나 데이터 구조를 수정하는 것을 방지한다.
- 베이스 - 한계 내부의 범위의 주소만 통과 가능하다.

---

**메모리 주소**

메모리에 저장된 정보의 위치는 주소로 나타낼 수 있다. 이때 주소는 물리 주소와 논리 주소로 나뉘어진다.

- **물리 주소(physical address)** : 정보가 실제로 저장된 하드웨어상의 주소(물리적 주소)
- **논리 주소(logical address)** : 실행 중인 프로그램 각각에 부여된 0번지부터 시작되는 주소 (가상 주소)
- CPU와 실행중인 프로그램은 현재 메모리 몇 번지에 무엇이 저장되어 있는지 알기 어렵다. 메모리에 저장된 정보는 시시각각 변하기 때문에 논리 주소를 사용한다.

---

**주소 바인딩**

주소 바인딩이란 프로그램의 논리 주소를 실제 메모리의 물리 주소로 매핑하는 작업이다. 크게 3가지가 존재한다.

- **컴파일 시간 바인딩(compile time binding)** : 컴파일 시간에 프로세스가 메모리에 상주할 위치를 미리 알 수 있다면 컴파일러는 절대코드를 생성할 수 있다. (논리주소와 물리주소가 일치하지만, 위치가 변경되면 코드를 다시 컴파일해야할 수 있다.)
- **적재 시간 바인딩(load time binding)** : 메모리 적재 위치를 컴파일 시점에 모르면 상대 주소를 생성한 다음 적재 시간에 시작 주소를 반영하여 사용자 코드 상의 주소를 재설정한다.
- **실행 시간 바인딩(run-time binding)** : 실행 시간에 주소를 지정한다. 프로세스가 수행 도중 다른 메모리 위치로 이동하는 경우, 주소 바인딩을 수행시간까지 연기해야한다. 이때 MMU라는 하드웨어의 도움이 필요하다.

---

**MMU(Memory Management Unit)**

<p align = "center">
   <img height = 300 width = 450 src = "https://github.com/user-attachments/assets/ff88109f-3dfd-429a-bfee-7ff389d7b23b" />
</p>

- MMU는 CPU가 발생시킨 논리 주소에 베이스 레지스터(relocation register, base register)값을 더하여 논리 주소를 물리 주소로 변환한다.
- 베이스 레지스터 : 프로그램의 가장 작은 물리 주소를 저장
- 논리 주소 : 프로그램의 시작점(base)로부터 떨어진 거리
